src/llama-sampling.cpp

```diff
@@ -1059,6 +1059,114 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
     };
 }
 
+// xtc
+
+struct llama_sampler_xtc {
+    const float    probability;
+    const float    threshold;
+    const float    threshold_max;
+    const size_t   min_keep;
+
+    const uint32_t seed;
+    uint32_t       seed_cur;
+
+    std::mt19937   rng;
+};
+
+static const char * llama_sampler_xtc_name(const struct llama_sampler * /*smpl*/) {
+    return "xtc";
+}
+
+static void llama_sample_xtc_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
+    auto * ctx = (llama_sampler_xtc *) smpl->ctx;
+
+    if (ctx->probability <= 0.0f
+        || ctx->threshold >= 1.0f
+        || ctx->threshold_max <= 0.0f
+        || ctx->threshold_max <= ctx->threshold
+        || cur_p->size <= 2) {
+        return;
+    }
+
+    std::uniform_real_distribution<float> distance(0.0f, 1.0f);
+    float chance = distance(ctx->rng);
+    if (chance > ctx->probability) return;
+
+    // in case it's not sorted/recalculated yet
+    llama_sampler_softmax_impl(cur_p);
+
+    int pos_first = -1;
+    int pos_last = 0;
+
+    for (size_t i = 0; i < cur_p->size; ++i) {
+        if (cur_p->data[i].p - ctx->threshold >= -1e-5) {
+            if (cur_p->data[i].p - ctx->threshold_max > 1e-3) pos_first = i;
+            pos_last = i;
+        } else {
+            break;
+        }
+    }
+
+    size_t to_remove = pos_last - (1 + pos_first);
+
+    if (to_remove < ctx->min_keep || to_remove < 1) return;
```

Comment:
This doesn't look like a correct way to respect `min_keep`. You need to clamp `to_remove` such that at the end `cur_p->size >= min_keep`.

Reply:
Yes, sorry, forgot to change this check.

Although, should I really clamp it or just check for `cur_p->size - to_remove` instead here? Clamping `to_remove` to a specific size sounds strange here. Min_p, for example, just doesn't apply its effect if `min_keep` requirement is not met.

Reply:
On that note, when is `min_keep` even used? It seems counterintuitive for many samplers, yet all of them follow it.

Reply:
It's something that has been there from the beginning, I am not sure that it is very useful.